cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfRobotComponent.cpp

```diff
@@ -0,0 +1,322 @@
+    // follow the root link's pose.
+    PrimaryComponentTick.bCanEverTick = true;
+    PrimaryComponentTick.TickGroup = ETickingGroup::TG_PostPhysics;
+
+    // UInputActionComponent
+    input_action_component_ = CreateDefaultSubobject<UInputActionComponent>(Unreal::toFName("input_action_component"));
+    SP_ASSERT(input_action_component_);
+    // Need to explicitly set this up so that the component hierarchy is well-defined.
+    input_action_component_->SetupAttachment(this);
+}
+
+UUrdfRobotComponent::~UUrdfRobotComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    // Objects created with CreateDefaultSubobject, DuplicateObject, LoadObject, NewObject don't need to be cleaned up explicitly.
+
+    LinkComponents.Empty();
+    JointComponents.Empty();
+
+    link_components_.clear();
+    joint_components_.clear();
+
+    SP_ASSERT(input_action_component_);
+    input_action_component_ = nullptr;
+}
+
+void UUrdfRobotComponent::BeginPlay()
+{
+    USceneComponent::BeginPlay();
+
+    // Get player input actions from the config system if it is initialized, otherwise use hard-coded keyboard actions, which
+    // can be useful for debugging.
+    std::map<std::string, std::map<std::string, std::vector<double>>> input_actions;
+    if (Config::s_initialized_) {
+        input_actions = Config::get<std::map<std::string, std::map<std::string, std::vector<double>>>>("URDF_ROBOT.URDF_ROBOT_COMPONENT.INPUT_ACTIONS");
+    } else {
+        input_actions = DEFAULT_INPUT_ACTIONS;
+    }
+
+    input_action_component_->bindInputActions(input_actions);
+    input_action_component_->apply_input_action_func_ = [this, input_actions](const std::string& key) -> void {
+        if (EnableKeyboardControl) {
+            // only assert if we're not in the editor
+            bool assert_if_action_is_inconsistent_with_joint = !WITH_EDITOR;
+            bool assert_if_joint_not_found = !WITH_EDITOR;
+            applyAction(input_actions.at(key), assert_if_joint_not_found, assert_if_action_is_inconsistent_with_joint);
+        }
+    };
+}
+
+void UUrdfRobotComponent::TickComponent(float DeltaTime, enum ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
+{
+    USceneComponent::TickComponent(DeltaTime, TickType, ThisTickFunction);
+
+    if (!ticked_once_) {
+        SP_LOG_CURRENT_FUNCTION();
+        initializeDeferred();
+        ticked_once_ = true;
+    }
+
+    // Update this component's pose to match root link component
+    SP_ASSERT(root_link_component_);
+    bool sweep = false;
+    FHitResult* hit_result = nullptr;
+    SetWorldLocationAndRotation(root_link_component_->GetComponentLocation(), root_link_component_->GetComponentRotation(), sweep, hit_result, ETeleportType::None);
+}
+
+std::map<std::string, ArrayDesc> UUrdfRobotComponent::getActionSpace() const
+{
+    std::map<std::string, ArrayDesc> action_space;
+
+    if (Std::contains(action_components_, "control_joints")) {
```

Comment:
```
    std::map<std::string, ArrayDesc> action_space;

    if (Std::contains(action_components_, "control_joints")) {
        for (auto joint_component : joint_components_) {
            action_space.merge(joint_component->getActionSpace();
        }
    }

    return action_space;
```

Reply:
Updated code

Reply:
I'm an idiot. This should be `auto&` because the iterator isn't a pointer type.

Reply:
I think an even cleaner code would be to iterate over JointComponents and LinkComponents array instead of the maps. In this case it's just `auto` :P

Reply:
Haha yes agreed, if we're iterating over `JointComponents` and `LinkComponents` we should use `auto`.